Add new documentation to explain how to map JKS to PEM files for ZooKeeper SSL (#8046)
DataDog/integrations-core: zk/README.md

After you have ZooKeeper set up with SSL, you can also configure the Datadog Agent to connect to ZooKeeper using SSL. If you already have authentication set up using JKS files, follow the steps below to convert them to PEM files for TLS/SSL configuration.

The following example commands assume that your JKS `truststore` and `keystore` files are called:

- `server_truststore.jks`
- `server_keystore.jks` 
- `client_truststore.jks`
- `client_keystore.jks`

The instructions also assume that both sides' `keystore` and `truststore` files have each other's certificates with aliases `server_cert` and `client_cert`, meaning that a Java ZooKeeper client can already connect to a ZooKeeper server.
If your private key has a password, make sure this password is included in the `config.yaml` file for config option `tls_private_key_password`.

To convert the JKS files to PEM files:

1. Get the `ca_cert.pem` file from `client_truststore.jks`, since the client's truststore contains the certificate of the server that is trustable:
    ```
    keytool -exportcert -file ca_cert.pem -keystore client_truststore.jks -alias server-cert -rfc
    ```
   

> 2. Get the `cert.pem` file from `server_truststore.jks`, since `cert.pem`:

Reviewer: ... either delete ", since cert.pem" or we are missing the end of the sentence.

Author: Oops good catch, will add to sentence.

    ```
    keytool -exportcert -file cert.pem -keystore server_truststore.jks -alias client-cert -rfc
    ```

3. Get the `private_key.pem` file from `client_keystore.jks`, since the client's `keystore` contains the private key of the client for alias `client-cert`:
    ```
    keytool -importkeystore -srckeystore client_keystore.jks -destkeystore private_key.p12 -srcstoretype jks -deststoretype pkcs12 -srcalias client-cert
    ```   

> 4. Run the `openssl pkcs12` command, which exports both the client `cert.pem` and the `private_key.pem` for the certificate, so the ZooKeeper integration can use them via the `tls_cert` option. Add `-nodes` to this command if you want to get a non-password-protected `private_key.pem` file:

Reviewer 2: You can remove step 2 and only use step4 instead. The openssl command you're using creates a pem file which contains both the private key and the certificate. i.e this file is the 'cert.pem' of the client, and no need to store the private_key into another file. The `tls_cert` config is able to read and parse pem files which contains both the cert and the private key.

2. [Restart the Agent][6].

#### Enabling SSL

Zookeeper 3.5 introduced the ability to use SSL authentication. The agent is able to connect to Zookeeper using SSL as well.

> One thing to note is that Zookeeper uses `truststores` and `keystores` JKS files while the Agent TLS/SSL configuration uses `tls_ca_cert`, `tls_cert`, and `tls_private_key` config options that accept PEM files.
> It may be confusing at first for those familiar with the JKS files to map to PEM files.
> 
> It is assumed that one already has the server and client truststore and keystore files to start Zookeeper in SSL mode in the first place.
> More information on setting up Zookeeper in SSL mode first can be found [here][13].
> 
> We will use following files as examples:

Reviewer: ```suggestion
```

- `server_truststore.jks`
- `server_keystore.jks` 
- `client_truststore.jks`
- `client_keystore.jks`

> We also assume that both sides' `keystore`s and `truststore`s have each other's certificates with alias `server_cert` and `client_cert`, meaning that a Java Zookeeper client can already connect to a Zookeeper server.

Reviewer: ```suggestion
The instructions also also assume that both sides' `keystore`and `truststore` files have each other's certificates with alias `server_cert` and `client_cert`, meaning that a Java Zookeeper client can already connect to a Zookeeper server.
```

> You may need to enter a password for these commands, in which case this password should be included in the `tls_private_key_password` config option.

Reviewer: ```suggestion
If you need to enter a password, add the option `-tls_private_key_password <PASSWORD>` to each command.
```

> 

Reviewer: ```suggestion
To convert the JKS files to PEM files:

```

> 1. We can get `ca_cert.pem` file from `client_truststore.jks`, since the client's truststore should contain the certificate of the server that is trustable.

Reviewer: ```suggestion
1. Get the `ca_cert.pem` file from `client_truststore.jks`, since the client's truststore contains the certificate of the server that is trustable:
```

    ```
    keytool -exportcert -file ca_cert.pem -keystore client_truststore.jks -alias server-cert -rfc
    ```
   

> 2. Next, we can get `cert.pem` file from `server_truststore.jks`, since `cert.pem` is the certificate of the client.

Reviewer: ```suggestion
2. Get the `cert.pem` file from `server_truststore.jks`, since `cert.pem`:
```

    ```
    keytool -exportcert -file cert.pem -keystore server_truststore.jks -alias client-cert -rfc
    ```

> 3. Finally, we will get `private_key.pem` file from `client_keystore.jks`, as the client's `keystore` contains the private key of the client for alias `client-cert`.

Reviewer: ```suggestion
3. Get the `private_key.pem` file from `client_keystore.jks`, since the client's `keystore` contains the private key of the client for alias `client-cert`:
```

    ```
    keytool -importkeystore -srckeystore client_keystore.jks -destkeystore private_key.p12 -srcstoretype jks -deststoretype pkcs12 -srcalias client-cert

>     openssl pkcs12 -in private_key.p12 -out private_key.pem
>     ```
> 
> Note: Adding `-nodes` to the `openssl` command is required here if you want to get a non-password protected `private_key.pem`.
> Note2: This exports both the client cert.pem and the `private_key.pem` for the certificate, which makes it easy for the integration to use via the `tls_cert` option.

Reviewer: Instead of having the two commands in the one step, with a bunch of notes, make this openssl command a separate step, something like this: 
```suggestion
4. Run the `openssl pkcs12` command, which exports both the client `cert.pem` and the `private_key.pem` for the certificate, so the Zookeeper integration can use them via the `tls_cert` option. Add `-nodes` to this command if you want to get a non-password-protected `private_key.pem` file:
    openssl pkcs12 -in private_key.p12 -out private_key.pem
 
```

You'll have to fix up the codeblock fences (three backticks) with this, too.

2. [Restart the Agent][6].

#### Enabling SSL

Zookeeper 3.5 introduced the ability to use SSL authentication. The agent is able to connect to Zookeeper using SSL as well.

One thing to note is that Zookeeper uses `truststores` and `keystores` JKS files while the Agent TLS/SSL configuration uses `tls_ca_cert`, `tls_cert`, and `tls_private_key` config options that accept PEM files.
It may be confusing at first for those familiar with the JKS files to map to PEM files.

It is assumed that one already has the server and client truststore and keystore files to start Zookeeper in SSL mode in the first place.
More information on setting up Zookeeper in SSL mode first can be found [here][13].

We will use following files as examples:
- `server_truststore.jks`
- `server_keystore.jks` 
- `client_truststore.jks`
- `client_keystore.jks`

We also assume that both sides' `keystore`s and `truststore`s have each other's certificates, meaning that a Java Zookeeper client can already connect to a Zookeeper server.

> _We first will get `cert.pem` file from `client_truststore.jks`._

Reviewer 2: That doesn't seem right. The client truststore contains the certificate of the server that is trustable. This maps to `ca_cert.pem`

1. Use `keytool` to export `client_truststore.jks` to a PKCS12 file called `cert.p12`:
    ```

>     keytool -importkeystore -srckeystore client_truststore.jks -destkeystore cert.p12 -srcstoretype jks -deststoretype pkcs12

Reviewer 2: Confused about the commands. From what I can read online, you need to use `-exportkeystore` to export the certificate to a cert file. And use `-importkeystore` to export the private key from a keystore.
I'm surprised all the commands are the same in here, can we double-check that?

Also when exporting the certs (both ca_cert and cert), couldn't we use the `rfc` flag of `exportkeystore` to squash those two operations into one?

Finally a keystore/truststore contains multiple entries, I assume we need to specify one using the `-alias` flag of the `exportkeystore` command.

Ref: https://docs.oracle.com/javase/8/docs/technotes/tools/unix/keytool.html#keytool_option_exportcert

Author: I couldn't find `-exportkeystore` in that doc, but saw `-exportcert`, is that what you meant?

I tested `keytool -importkeystore` with `openssl pkcs12` and `keytool -exportcert -rfc` for the `ca_cert` and `cert` files, and I got the same results. However, I'll change it to use `-exportcert -rfc` instead since it is one fewer step like you said.

Reviewer 2: Yup sorry, I meant exportcert

    ```
2. Use `openssl` to export the new PKCS12 `cert.p12` file to `cert.pem` PEM file:
    ```
    openssl pkcs12 -in cert.p12 -out cert.pem
    ```

> _Next we will get `ca_cert.pem` file from `server_truststore.jks`._

Reviewer 2: The server_truststore contains the certificate of the client that is trusted. This maps to `cert.pem`, i.e the cert of the client.

1. Use `keytool` to export `server_truststore.jks` to a PKCS12 file called `ca_cert.p12`:

    ```
    keytool -importkeystore -srckeystore server_truststore.jks -destkeystore ca_cert.p12 -srcstoretype jks -deststoretype pkcs12
    ```
2. Use `openssl` to export the new PKCS12 `ca_cert.p12` file to `ca_cert.pem` PEM file:
    ```
    openssl pkcs12 -in ca_cert.p12 -out ca_cert.pem
    ```

_Finally we will get `private_key.pem` file from `client_keystore.jks`._
1. Use `keytool` to export `client_keystore.jks` to a PKCS12 file called `private_key.p12`:
    ```

>     keytool -importkeystore -srckeystore client_keystore.jks -destkeystore private_key.p12 -srcstoretype jks -deststoretype pkcs12

Reviewer 2: I believe here it makes sense to use `importkeystore` as that command allows you to import an existing keystore and convert it to another format. But one should specify which entries of the keystore needs to be extracted using `-srcalias`

Author: Good catch, it is true that there can be multiple entries.

    ```
   
2. Use `openssl` to export the new PKCS12 `private_key.p12` file to `private_key.pem` PEM file:
    ```

>     openssl pkcs12 -in private_key.p12 -out private_key.pem

Reviewer 2: FTR: Adding `-nodes` is required here if you want to get a non-password protected private_key.pem.

Note2: This exports both the client cert.pem and the private_key for the certificate, which makes it easy for the integration to use via the `tls_cert` option.

Author: I will add this as a note.
